### PR TITLE
Improve Dolby Vision detection for Android

### LINKED ISSUE #20740: Dolby Vision files are not recognized if embed in a mkv container
### Describe the bug
Dolby Vision files are not recognized if embed in a mkv container and played as SDR files.

You can download this small file to reproduce the issue: https://send.portailpro.net/download/e34c450ba59cb290/#QThGw5hqmfE1WL6584Nl1w
## Expected Behavior
Here is a clear and concise description of what was expected to happen:

Dolby Vision should be recognized if embed in a mkv container
## Actual Behavior

Dolby Vision files embed in a mkv container are played as SDR files.
## Possible Fix

I tried to extract only the video stream with mkvtoolnix but no luck.
### To Reproduce

1. Download the file : https://send.portailpro.net/download/e34c450ba59cb290/#QThGw5hqmfE1WL6584Nl1w
2. Play.
3.

### Debuglog

The debuglog can be found here:

### Screenshots 
Here are some links or screenshots to help explain the problem:

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

### Your Environment
Used Operating system:

 - [x] Android
 - [ ] iOS
 - [ ] tvOS
 - [ ] Linux
 - [ ] OSX
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name: Android 9
 - Kodi version: 20.x (a december 2021 release)

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

### PR DESCRIPTION
**Pull request by sscobici**
## Description
Improve Dolby Vision detection on Android devices.
Possibly fixes: https://github.com/xbmc/xbmc/issues/20740
Probably substitutes https://github.com/xbmc/xbmc/pull/22410

## Motivation and context
Some DV files are played as HDR or SDR instead of Dolby Vision. This PR is intended to fix this.

## How has this been tested?
I've tested this on Ugoos AM6 Plus (firmware 0.5.4) and LG C2 TV with the sample files from Kodi Samples 4k https://kodi.wiki/view/Samples:
```
24. http://media.developer.dolby.com/DolbyVision_Atmos/mp4/P81_GlassBlowing2_3840x2160%4059.94fps_15200kbps_fmp4.mp4
HDR format : Dolby Vision, Version 1.0, dvhe.08.09, BL+RPU, HDR10 compatible / SMPTE ST 2086, HDR10 compatible
Codec ID : hev1

13. https://drive.google.com/file/d/1cTRNwacsV-8J8PtcYNVDXBmGZF1SOynj/view?usp=sharing iPhone11_4K-recorder_59.940DV+HLG.mov
HDR format : Dolby Vision, Version 1.0, dvhe.08.12, BL+RPU, HLG compatible
Codec ID : hvc1

19. https://img.photographyblog.com/reviews/apple_iphone_13_pro/sample_images/4K60p.mov
HDR format : Dolby Vision, Version 1.0, dvhe.08.10, BL+RPU, HLG compatible
Codec ID : hvc1
```

Files should be played from the local drive, Kodi have some cache issues when playing them from http storage because they are small, other big DV files play normally from remote storage.

It should also help with these files but those are DV FEL and have freeze issues even if they are detected as DV on TV.
```
21. https://drive.google.com/file/d/0BwxFVkl63-lEWmNWcGl4eVRWS2M/view?usp=sharing&resourcekey=0-WkmICjhK5yus4EAaPuSuSw LG Demo DolbyVision Trailer.mkv
HDR format : Dolby Vision, Version 1.0, dvhe.04.06, BL+EL+RPU
Codec ID : 36

20. https://drive.google.com/file/d/0BwxFVkl63-lEc3k0aldZaGtaVm8/view?resourcekey=0-ECypkFcgdKx1YiojpUCWBA LG Demo DolbyVision Comparison.mkv
HDR format : Dolby Vision, Version 1.0, dvhe.04.06, BL+EL+RPU
Codec ID : 36
```

## What is the effect on users?
The end user will be able to watch more Dolby Vision files in DV format on Android.

## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [x] **Improvement** (non-breaking change which improves existing functionality)

## Checklist:

- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by thexai** [comment]
Passing `hdrType` seems like a good idea to me, this way it is not necessary to patch the existing `codec_tag` and it seems that more cases can be covered. It also looks better for future expansions.

**Review by thexai — APPROVED** [review]
Runtime tested Shield Pro 2019 (FW 9.1.1)

Tested many files: mkv, mp4, various profiles 4, 5, 7.6, 8, 8.1.... 
tested files with DV only and files with DV + HDR10 fallback and DV + SDR fallback 

Absolutely all samples played fine (correct colors) and in DV mode.

Seems all is good enough 👍

EDIT: It is foreseeable that some caching related issue arises but in any case has not direct relation with this PR and this is v21 pre alpha.

File: xbmc/cores/VideoPlayer/DVDStreamInfo.h [hunk 1]
```diff
@@ -68,6 +68,7 @@ class CDVDStreamInfo
   int orientation; // orientation of the video in degrees counter clockwise
   int bitsperpixel;
   int bitdepth;
+  StreamHdrType hdrType;
```
[the remark this answers is not in the document]
Comment (sscobici) [reply xbmc/cores/VideoPlayer/DVDStreamInfo.h:71]: Do you want me to refactor all file and apply this apply this convention to all variabile names to be consistent? That will be a big change.

[the remark this answers is not in the document]
Comment (thexai) [reply xbmc/cores/VideoPlayer/DVDStreamInfo.h:71]: That doesn't make much sense here. In any case, these types of changes should be made in a separate PR.

**Comment by thexai** [comment]
Thank you for your first contribution.

**Comment by jjd-uk** [comment]
> I'm sorry if I'm being stupid, but where can I download a version of kodi including this life saving fix? Please and thank you.

From http://mirrors.kodi.tv/nightlies/android/ but you'll need to wait until the 11th Jan when the next Nightly build will appear.

**Comment by guilamu** [comment]
> From http://mirrors.kodi.tv/nightlies/android/ but you'll need to wait until the 11th Jan when the next Nightly build will appear.

Thank you very much. All versions (Nexus, Matrix, Omega, Leia) will have the fix included?

**Comment by Nuklear92** [comment]
> > From http://mirrors.kodi.tv/nightlies/android/ but you'll need to wait until the 11th Jan when the next Nightly build will appear.
> 
> 
> 
> Thank you very much. All versions (Nexus, Matrix, Omega, Leia) will have the fix included?
> 
> 

I believe this is only for the Omega v21.0 that they're currently working on. But the great thing is that Kodi finally supports all DV vision profiles and containers. With the next version of Kodi after Nexus. 😃

**Comment by guilamu** [comment]
> > > From http://mirrors.kodi.tv/nightlies/android/ but you'll need to wait until the 11th Jan when the next Nightly build will appear.
> > 
> > 
> > Thank you very much. All versions (Nexus, Matrix, Omega, Leia) will have the fix included?
> 
> I believe this is only for the Omega v21.0 that they're currently working on. But the great thing is that Kodi finally supports all DV vision profiles and containers. With the next version of Kodi after Nexus. 😃

Oh God, no... it means this fix won't be on a stable release before months or years, right?

**Comment by guilamu** [comment]
> Maybe, maybe not. I hear that the next Kodi version 21 won't take as long to be released, as it's been with previous major releases. Let's hope that by end of this year we have Omega almost ready for public. 🤞

All right, I guess I'll be stuck with Maven's releases for a few more months. Anyhow, thanks for your answers.

**Comment by Nuklear92** [comment]
> > Maybe, maybe not. I hear that the next Kodi version 21 won't take as long to be released, as it's been with previous major releases. Let's hope that by end of this year we have Omega almost ready for public. 🤞
> 
> 
> 
> All right, I guess I'll be stuck with Maven's releases for a few more months. Anyhow, thanks for your answers.

You welcome, yeah I'll suggest sticking to that until v21 start going into the public betas at least and most of the addons start supporting Omega. Good thing is we finally have full support for Dolby Vision and no need to rely on external players or special builds for this with the next major Kodi version. 👍

File: xbmc/cores/VideoPlayer/DVDStreamInfo.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (lrusak) [reply xbmc/cores/VideoPlayer/DVDStreamInfo.h:71]: We have guidelines for a reason. An entire refactor isn't needed. I only asked for this single new member.

**Comment by jjd-uk** [comment]
V20 Nexus wasn't branched from Master that long ago, so the Nighlies based on Master (now for v21) won't have deviated too much yet so should still be pretty stable.

**Comment by KyleSanderson** [comment]
> V20 Nexus wasn't branched from Master that long ago, so the Nightlies based on Master (now for v21) won't have deviated too much yet so should still be pretty stable.

Right, but you won't have any security patches or similar on the stable platform. This (obviously) needs to be backported, otherwise it's the same situation where people need to use the KodiNerds release until Kodi catches up in 2024+.

**Comment by sscobici** [comment]
This fix depends on FFmpeg 5.2.1 which was merged just 3 days ago https://github.com/xbmc/xbmc/pull/21248

**Comment by jjd-uk** [comment]
And there's zero change of FFmpeg 5.2.1 being backported for v20 Nexus

**Comment by KyleSanderson** [comment]
> And there's zero chance of FFmpeg 5.2.1 being backported for v20 Nexus

sure, then merge this into the stable branch. https://github.com/xbmc/FFmpeg/pull/24

**Comment by HitcherUK** [comment]
For those interested here's the first release -

http://mirrors.kodi.tv/nightlies/android/arm64-v8a/master/kodi-20230111-1f9640a3-master-25562-arm64-v8a.apk

Tested myself and DV works perfectly on my Shield TV Pro.

**Comment by monsoons** [comment]
@sscobici Thank you for your PR. Great job!
I've tested some profile 7 DV mkvs(UHD rips) with Ugoos am6b plus and also experienced unstable playback(video freezing, audio drops, etc...).

As you know, unlike shield TV pro which ignores EL, Ugoos am6b plus can decode 2160p and 1080p stream simultaneously and the playback is smooth in other media player such as plex and novaplayer.

I guess something goes wrong in the process of decoding EL in Kodi. Do you think this issue is fixable?

**Comment by sscobici** [comment]
I still trying to understand what is the root cause.

**Comment by quietvoid** [comment]
> I am still trying to understand what is the root cause.

@sscobici You could test this out: https://github.com/xbmc/xbmc/pull/22546
It might help for profile 7, but the EL will ignored.

**Comment by sscobici** [comment]
@quietvoid I created the forum thread for further discussion: https://forum.kodi.tv/showthread.php?tid=371521 see you there.
Device can correctly play profile 7 FEL with Nova player, at least with the sample that have consistent issues in Kodi.
So there SHOULD be some difference but I still cannot find it.

**Comment by TwillightMercy** [comment]
Hi question,

If I compile myself Android APK of Kodi 20.0 Nexus branch + this commit inserted manually in the code - Dolby vision support should work?
I tried and I don't get DV notification.

**Comment by sscobici** [comment]
In Nexus, ffmpeg doesn't detect DV that's why this PR doesn't work there. For Nexus branch you need https://github.com/xbmc/FFmpeg/pull/24 then this PR will work for mkv

**Comment by TwillightMercy** [comment]
> In Nexus, ffmpeg doesn't detect DV that's why this PR doesn't work there. For Nexus branch you need [xbmc/FFmpeg#24](https://github.com/xbmc/FFmpeg/pull/24) then this PR will work for mkv

Thanks for the reply.

What about this?
"This fix depends on FFmpeg 5.2.1 which was merged just 3 days ago https://github.com/xbmc/xbmc/pull/21248"

**Comment by guilamu** [comment]
http://mirrors.kodi.tv/nightlies/android/arm64-v8a/master/kodi-20230111-1f9640a3-master-25562-arm64-v8a.apk

This one works perfectly fine with dv on Nvidia shield TV.

**Comment by TwillightMercy** [comment]
> android/arm64-v8a/master/kodi-20230111-1f9640a3-master-25562-arm64-v8a.apk

Thank you, I'm looking for make DV work on Nexus branch, because I want to stay on 20.0 stable release, but with DV support.

Will backporting all dependencies to Nexus branch happen? 🙏

**Comment by sscobici** [comment]
> What about this? "This fix depends on FFmpeg 5.2.1 which was merged just 3 days ago #21248"

Please the description of PR https://github.com/xbmc/FFmpeg/pull/24. It contains only the code from ffmpeg 5 that is required for DV detection in mkv.

**Comment by TwillightMercy** [comment]
> > What about this? "This fix depends on FFmpeg 5.2.1 which was merged just 3 days ago #21248"
> 
> Please read the description of PR [xbmc/FFmpeg#24](https://github.com/xbmc/FFmpeg/pull/24). It contains only the code from ffmpeg 5 that is required for DV detection in mkv.

Thank you I'll try. but anyway do you know about backporting FFmpeg + this PR to Nexus branch?

**Comment by TwillightMercy** [comment]
> > What about this? "This fix depends on FFmpeg 5.2.1 which was merged just 3 days ago #21248"
> 
> Please read the description of PR [xbmc/FFmpeg#24](https://github.com/xbmc/FFmpeg/pull/24). It contains only the code from ffmpeg 5 that is required for DV detection in mkv.

Thank you i've read it. so the other PR I sent is already merged to Nexus?

because there is a comment in https://github.com/xbmc/FFmpeg/pull/24:
"since we're probably not going to bump to ffmpeg 5 for Nexus, this looks like a pretty isolated change"

**Comment by TwillightMercy** [comment]
> In Nexus, ffmpeg doesn't detect DV that's why this PR doesn't work there. For Nexus branch you need [xbmc/FFmpeg#24](https://github.com/xbmc/FFmpeg/pull/24) then this PR will work for mkv

Also, this PR https://github.com/xbmc/FFmpeg/pull/24 is for FFmpeg repo not for Kodi, I'm not sure to how to use it to make this PR work on Nexus branch

**Comment by sscobici** [comment]
It can be implemented as a patch to ffmpeg 4 in kodi repo

**Comment by the-black-eagle** [comment]
@TwillightMercy  No, stuff won't be backported to Nexus.

To get V20 (Nexus) to work with DV, you need to patch FFmpeg that Kodi builds against with https://github.com/xbmc/FFmpeg/pull/24 Then build Kodi using that patched FFmpeg and DV will work in v20.

**Comment by TwillightMercy** [comment]
> @TwillightMercy No, stuff won't be backported to Nexus.
> 
> To get V20 (Nexus) to work with DV, you need to patch FFmpeg that Kodi builds against with [xbmc/FFmpeg#24](https://github.com/xbmc/FFmpeg/pull/24) Then build Kodi using that patched FFmpeg and DV will work in v20.

Thanks for the reply!

I've cloned Nexus branch with this PR code fix, but i'm not sure how to do patch FFmpeg that Kodi builds, it's a different repository and I'm building Kodi with xmbc/xbmc repo. Can you explain to me what I need to do?

**Comment by TwillightMercy** [comment]
> It can be implemented as a patch to ffmpeg 4 in kodi repo

Can you instruct me how 🙏
I know how to apply this current PR code to kodi branch, but ffmpeg is different repo, not sure what to do

**Comment by fuzzard** [comment]
Can you please stop spamming a bunch of different prs/issues with support questions

**Comment by Moxh3** [comment]
Tested 21 Omega Alpha 1 on FireTV Cube 3rd Gen streaming Smile (DV) through Jellyfin 10.8.10 in MKV format. Still only showed the movie in HDR10 and no DV.

**Pull request merged**